namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala

```diff
@@ -0,0 +1,72 @@
+package io.buoyant.namer.dnssrv
+
+import java.net.InetSocketAddress
+
+import com.twitter.finagle._
+import com.twitter.finagle.stats.StatsReceiver
+import com.twitter.logging.Logger
+import com.twitter.util.Activity.State
+import com.twitter.util._
+import org.xbill.DNS
+
+class DnsSrvNamer(prefix: Path, resolver: DNS.Resolver, timer: Timer, refreshInterval: Duration, stats: StatsReceiver)
+  extends Namer {
+
+  override def lookup(path: Path): Activity[NameTree[Name]] = memoizedLookup(path)
+
+  private val log = Logger.get("dnssrv")
+  private val memoizedLookup: (Path) => Activity[NameTree[Name]] = Memoize {
+    case path@Path.Utf8(address, _) =>
+      val id = path.take(1)
+      Activity(Var.async[State[NameTree[Name]]](Activity.Pending) { state =>
+        timer.schedule(refreshInterval) {
+          val next = lookupSrv(address, prefix ++ id) match {
+            case Return(nameTree) => Activity.Ok(nameTree)
+            case Throw(e) => Activity.Failed(e)
+          }
+          state.update(next)
+        }
+      })
+    case _ => Activity.value(NameTree.Neg)
+  }
+
+  private[dnssrv] def lookupSrv(address: String, id: Path): Try[NameTree[Name]] = Try {
+    val question = DNS.Record.newRecord(
+      DNS.Name.fromString(address),
+      DNS.Type.SRV,
+      DNS.DClass.IN
+    )
+    val query = DNS.Message.newQuery(question)
+    log.debug("looking up %s", address)
+    val m = resolver.send(query)
+    log.debug("got response %s", address)
+    m.getRcode match {
+      case DNS.Rcode.NXDOMAIN =>
+        log.trace("no results for %s", address)
+        NameTree.Neg
+      case DNS.Rcode.NOERROR =>
+        val hosts = m.getSectionArray(DNS.Section.ADDITIONAL).collect {
+          case a: DNS.ARecord => a.getName -> a.getAddress
+        }.toMap
+        val srvRecords = m.getSectionArray(DNS.Section.ANSWER).collect {
+          case srv: DNS.SRVRecord =>
+            hosts.get(srv.getTarget) match {
+              case Some(inetAddress) => Address(new InetSocketAddress(inetAddress, srv.getPort))
+              case None => Address(srv.getTarget.toString, srv.getPort)
+            }
+        }
+        if (srvRecords.isEmpty) {
+          // valid DNS entry, but no instances.
+          // for some reason, NameTree.Empty doesn't work right
+          log.trace("empty response for %s", address)
+          NameTree.Neg
```

Comment:
It's debatable if you want NameTree.Neg here or a leaf with Addr.Bound containing an empty set.  The former will consider the name invalid and will fall back to any alternatives in the NameTree.  The later will be considered a valid name but any requests will fail because there are no addresses to send to.

Reply:
a major linkerd use-case for us is to load-balance across two or more deployments of a service (with different SRV records), i.e. the dtab looks like this:
```
    /svc/myservice =>
               /dnssrv/myservice1.srv.example.org &
               /dnssrv/myservice2.srv.example.org;
```
if `myservice1` returns `Addr.Bound(Set.empty)`, would linkerd still call `myservice2`?

Reply:
If you use an empty address set then half of the time you'll pick the first branch of the union and fail the request and half of the time you'll pick the second branch of the union and use myservice2

If you use NameTree.Neg then you'll skip to the myservice2 every time.

Reply:
sounds like `NameTree.Neg` is the right choice then :). In general, why would linkerd _not_ fallback/load-balance for empty sets? I see this has come up in #1612 and #1549, too.